[JIT] Support calling Tensor.element_size() in TorchScript

xush6528 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#33808 [JIT] Support calling Tensor.element_size() in TorchScript**

# Problem

Resolves https://github.com/pytorch/pytorch/issues/33620

Differential Revision: [D7050644](https://our.internmc.facebook.com/intern/diff/D7050644/)

xush6528 committed:
[WIP] [JIT] Support calling Tensor.element_size() in TorchScript

# Problem

https://github.com/pytorch/pytorch/issues/33620

Differential Revision: [D7050644](https://our.internmc.facebook.com/intern/diff/D7050644/)

[ghstack-poisoned]

eellison commented on tools/jit/gen_jit_dispatch.py:
```diff
@@ -404,6 +404,7 @@ def declkey(decl):
         ('strides', 'aten::strides(Tensor self) -> int'),
         ('dim', 'aten::dim(Tensor self) -> int'),
         ('numel', 'aten::numel(Tensor self) -> int'),
+        ('element_size', 'aten::element_size(Tensor self) -> int'),
```
@bhosmer, is there where we should be putting all of the fast-path functions ? https://github.com/pytorch/pytorch/issues/30774

bhosmer replied:
Yep, I don't know the history but this is the current collecting spot AFAIK

bhosmer replied:
Re #30774 yeah, this adds element_size to the list - I don't know that it's a tipping point though - and it might be a bit before the unification happens. What's your take though?

eellison replied:
Yea i think it's fine to add it. It might be nice in some future time to make this list more discoverable (I didn't know about it) via native declaration, but I don't think that should block increasing parity in the meantime.

eellison commented on test/jit/test_tensor.py:
```diff
@@ -0,0 +1,29 @@
+
+    def test_element_size(self):
+        @torch.jit.script
+        def get_element_size_script(x):
```
not really sure this warrants it's own file... we have other tests on tensor methods in test_jit.py

xush6528 replied:
Moved to test_jit.py

eellison commented on c10/core/TensorImpl.h:
```diff
@@ -408,6 +408,10 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
     return numel_;
   }
 
+  virtual int64_t element_size() const {
```
Why virtual ?

xush6528 replied:
Removed.

eellison reviewed (commented):
LGTM, requesting review from @bhosmer or @yf225 before accepting tho. Not sure if this is the right place to bind this (although it looks to me that way).

eellison commented on c10/core/TensorImpl.h:
```diff
@@ -408,6 +408,10 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
   }
 
+  virtual int64_t element_size() const {
+    return static_cast<int64_t>(storage_.itemsize());
```
element_size is already defined on `TensorBody.h`, why do we need to duplicate it?

xush6528 replied:
Removed.

yf225 commented on c10/core/TensorImpl.h:
```diff
@@ -408,6 +408,10 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
     return numel_;
   }
 
+  virtual int64_t element_size() const {
```
Why do we need to add this function? (Does TorchScript `Tensor.element_size()` directly bind to functions in TensorImpl?) We have `Tensor::element_size()` defined in `aten/src/ATen/templates/TensorBody.h`, is there a way to use that instead?

xush6528 replied:
@yf225 

The issue describes this missing parity, #33620.
Digging where the scripting error comes from, I noticed it comes from 

https://github.com/pytorch/pytorch/blob/ee28831341c0398b0e2a0791ffd77263217f0bff/torch/csrc/jit/script/sugared_value.cpp#L151

While, the similar `Tensor().numel()` works by running into this line. 
https://github.com/pytorch/pytorch/blob/ee28831341c0398b0e2a0791ffd77263217f0bff/torch/csrc/jit/script/sugared_value.cpp#L145

So I search for where `aten::numel` is registered and noticed that `tensor_impl_methods` list I am touching. And it says it works by adding methods in `TensorImpl` class.,

xush6528 commented:
TODO: Add method hint like https://github.com/pytorch/pytorch/pull/27294/.

xush6528 committed:
Update on "[WIP] [JIT] Support calling Tensor.element_size() in TorchScript"


# Problem

Resolves https://github.com/pytorch/pytorch/issues/33620

Differential Revision: [D7050644](https://our.internmc.facebook.com/intern/diff/D7050644/)

[ghstack-poisoned]

xush6528 committed:
Update on "[WIP] [JIT] Support calling Tensor.element_size() in TorchScript"


# Problem

Resolves https://github.com/pytorch/pytorch/issues/33620

Differential Revision: [D7050644](https://our.internmc.facebook.com/intern/diff/D7050644/)

[ghstack-poisoned]

eellison reviewed (approved):
LGTM if tests pass.

xush6528 committed:
Update on "[JIT] Support calling Tensor.element_size() in TorchScript"


# Problem

Resolves https://github.com/pytorch/pytorch/issues/33620

Differential Revision: [D7050644](https://our.internmc.facebook.com/intern/diff/D7050644/)

[ghstack-poisoned]
